Automatic filters from json add tests

fbudin69500 opened the pull request.

fbudin69500 committed:
ENH: Creating filters for DREAM3D based on JSON description files

fbudin69500 committed:
ENH: Addition of 'dim_vec' and 'itk_type' from members

Members with 'vector' type components such as itk::Point, itk::Size,
itk::Spacing, can be used and set in the GUI of the DREAM3D plugins.

fbudin69500 committed:
ENH: Adding automatically generated tests

Creating automatically generated unit tests for ITK-based
DREAM.3D filters that are automatically generated from JSON
files from the SimpleITK project.

fbudin69500 committed:
ENH: Adding automatically generated filters and corresponding tests

DREAM3D filters have been automatically generated from ITK ImageToImage filters.

fbudin69500 committed:
BUG: StaticCastScalar() and CastStdToVec3() require the same template arguments

To simplify the automatic filter creation, StaticCastScalar() and CastStdToVec3()
require the same template arguments.

fbudin69500 committed:
COMP: Remove compilation warnings

Specify methods that will override base class methods to avoid compilation warning
messages. Some warning messages are still visible due to commit [1].

[1] https://github.com/BlueQuartzSoftware/SIMPL/commit/405fc1a6e1246e1b9112f7614e16ae44b6a4fc0b

fbudin69500 committed:
ENH: Add information to automatically generated filters based on JSON

Description of the filter and of the parameters are extracted from the JSON
file and inserted in the markdown file providing help for the filters
that are automatically generated.

fbudin69500 committed:
ENH: Create base class for automatically generated tests

Automatically generated tests are now all derived from the base class
ITKTestBase.

thewtex commented on ITKImageProcessingFilters/ITKBilateralImage.cpp:
```diff
@@ -153,3 +153,10 @@ AbstractFilter::Pointer ITKBilateralImage::newFilterInstance(bool copyFilterPara
+//
+// -----------------------------------------------------------------------------
+const QString ITKBilateralImage::getSubGroupName()
+{ return "ImageEnhancement"; }
```
ITKImageEnhancement?

thewtex commented on ITKImageProcessingFilters/ITKIntensityWindowingImage.cpp:
```diff
@@ -156,3 +156,10 @@ AbstractFilter::Pointer ITKIntensityWindowingImage::newFilterInstance(bool copyF
+//
+// -----------------------------------------------------------------------------
+const QString ITKIntensityWindowingImage::getSubGroupName()
+{ return "IntensityImageFilters  MultiThreaded"; }
```
ITKImageIntensity?

thewtex commented on Utilities/SimpleITKJSONDream3DFilterCreation.py:
```diff
@@ -665,6 +666,18 @@ def GetDREAM3DParameterDescription(filter_member):
+                if '\ingroup' in line:
+                    split_line=line.split('\ingroup')
+                    return split_line[1].strip()
+    return "ITKNoModule"
```
May also need a check for starting with "ITK".

fbudin69500 replied:
Thanks for the review!

fbudin69500 committed:
ENH: ITK filters subGroupName is ITK module name

subGroupName for automatically generated ITK filters is
the name of the ITK module the ITK filter belongs too.

fbudin69500 committed:
ENH: Automatically generated filters can have different input and output pixel types

If output_pixel_type is specified in the JSON file describing the ITK filter,
it is used to specify the output pixel type in the DREAM3D filter automatically
generated.

fbudin69500 committed:
ENH: Adding custom_itk_cast to keywords that python script handles

fbudin69500 committed:
ENH: Using 'template_code_filename' and 'template_test_filename' JSON information

'template_code_filename' and 'template_test_filename' select the template that is used
to create the filter. Different ITK filters require different templates.

fbudin69500 committed:
ENH: Adding support for 'output_image_type'

Filters outputing images with a specific output image type described
in the JSON file can be automatically converted to DREAM.3D filters.

fbudin69500 committed:
BUG: Limit number of CPU cores used on CircleCI

When running 'ctest -D Experimental' to compile and test this plugin,
ctest compiles the plugin using all the cores that are available on
the machine, which increases the amount of memory used by the compiler,
and the compilation fails due to requiring to much memory.
To avoid this problem, and without completely re-writing the testing
system and using a CTest script, we just compile DREAM3D firsti, limiting
the number of cores requested for the compilation, and then run
'ctest -D Experimental'.

fbudin69500 committed:
ENH: Supporting RBGA and Vector images

fbudin69500 committed:
ENH: Add test for image component size.

fbudin69500 committed:
ENH: Replace explicit instantiation of fully specialized static method with overloaded templated function.

fbudin69500 committed:
ENH: Support tests for RGBA and Vector images

fbudin69500 committed:
ENH: Add include files in tests in case special DREAM3D data types are required.

fbudin69500 committed:
ENH: Separate supported read and write formats.

The list of supported read formats and write formats might be different.

fbudin69500 committed:
BUG: Updating filters to use latest version of SIMPL (12/8/2016)

SIMPL enum declarations have been modified [1] and required some code
update in the filters' source code to compile.

[1] https://github.com/BlueQuartzSoftware/SIMPL/commit/645ec07243e18c082017ca99921bb031bca571b6

fbudin69500 committed:
ENH: Removing ITKSaltAndPepperNoiseImage test

fbudin69500 committed:
COMP: Hiding warnings due to inconsistent-missing-override

The function "IsA()" is not declared as 'override' in SIMPLibSetGetMacros.h
in SIMPL [1]. The creates a lot of warnings when compiling ITKImageProcessing.
Following the example of DREAM3D, this commit hides these warnings.

[1] https://github.com/BlueQuartzSoftware/SIMPL/commit/405fc1a6e1246e1b9112f7614e16ae44b6a4fc0b
